lib/improver/nbhood.py

```diff
@@ -55,107 +59,242 @@ class BasicNeighbourhoodProcessing(object):
         self.unweighted_mode = bool(unweighted_mode)
 
     def __str__(self):
-        result = ('<NeighbourhoodProcessing: radius_in_km: {};' +
+        result = ('<NeighbourhoodProcessing: radii_in_km: {};' +
                   'unweighted_mode: {}>')
         return result.format(
-            self.radius_in_km, self.unweighted_mode)
+            self.radii_in_km, self.unweighted_mode)
+
+    def find_required_lead_times(self, cube):
+        """
+        Determine the lead times within a cube, either by reading the
+        forecast_period coordinate, or by calculating the difference between
+        the time and the forecast_reference_time.
+
+        Parameters
+        ----------
+        cube : Iris.cube.Cube
+            Cube from which the lead times will be determined.
+
+        Returns
+        -------
+        required_lead_times : Numpy array
+            Array containing the lead times, at which the radii need to be
+            calculated.
+
+        """
+        if cube.coords("forecast_period"):
+            required_lead_times = cube.coord("forecast_period").points
+        else:
+            if cube.coords("time") and cube.coords("forecast_reference_time"):
+                required_lead_times = (
+                    cube.coord("time").points -
+                    cube.coord("forecast_reference_time").points)
+            else:
+                msg = ("The forecast period coordinate is not available "
+                       "within {}."
+                       "The time coordinate and forecast_reference_time "
+                       "coordinate were also not available for calculating "
+                       "the forecast_period.".format(cube))
+                raise CoordinateNotFoundError(msg)
+        return required_lead_times
 
-    def get_grid_x_y_kernel_ranges(self, cube):
-        """Return grid cell numbers east and north for the kernel."""
+    def get_grid_x_y_kernel_ranges(self, cube, radii_in_km):
+        """
+        Return the number of grid cells in the x and y direction
+        to be used to create the kernel.
+
+        Parameters
+        ----------
+        cube : Iris.cube.Cube
+            Cube containing the x and y coordinates, which will be used for
+            calculating the number of grid cells in the x and y direction,
+            which equates to the size of the desired radii.
+        radii_in_km : Float
+            Radius in kilometres for use in specifying the number of
+            grid cells used to create a kernel.
+
+        Returns
+        -------
+        grid_cells_x : Integer
+            Number of grid cells in the x direction to be used to create the
+            kernel.
+        grid_cells_y : Integer
+            Number of grid cells in the y direction to be used to create the
+            kernel.
+
+        """
         try:
             x_coord = cube.coord("projection_x_coordinate").copy()
             y_coord = cube.coord("projection_y_coordinate").copy()
```

Comment:
Yes the neighbourhood processing will need to deal with non-regular grids, but as Gavin says, not in this request. For the neighbourhood radius definition, it is the distance in orthogonal directions from the central grid "point" and therefore the projection shouldn't affect the distance defined, but there may need to be some extra geometry when doing the neighbourhood processing to account for stretching in some directions. Definitely worth raising an issue in future- cheers.